File: frontend/src/services/webrtcService.js
/**
 * webrtcService.js
 *
 * Manages WebRTC PeerConnection and local media streams.
 * Signaling is delegated entirely to signalingService (backend-routed).
 */

import { sendSignal } from './signalingService';

const STUN_URL = import.meta.env.VITE_STUN_URL || 'stun:stun.l.google.com:19302';

const ICE_SERVERS = [{ urls: STUN_URL }];

let peerConnection = null;
let localStream = null;
let sessionToken = null;
let onRemoteStreamCallback = null;
let onIceConnectionStateCallback = null;

/**
 * Initialise local media (video + audio, or audio-only).
 * Returns the acquired MediaStream.
 */
export async function initLocalStream() {
  // Try video + audio first
  try {
    localStream = await navigator.mediaDevices.getUserMedia({ video: true, audio: true });
    return { stream: localStream, videoEnabled: true, audioEnabled: true };
  } catch (videoErr) {
    console.warn('Camera unavailable, falling back to audio-only:', videoErr);
  }

  // Fallback: audio only
  try {
    localStream = await navigator.mediaDevices.getUserMedia({ video: false, audio: true });
    return { stream: localStream, videoEnabled: false, audioEnabled: true };
  } catch (audioErr) {
    console.warn('Microphone also unavailable — receive-only mode');
  }

  // Final fallback: receive-only (no local tracks)
  localStream = null;
  return { stream: null, videoEnabled: false, audioEnabled: false };
}

/** Create a new RTCPeerConnection and attach local tracks */
function createPeerConnection(token, onRemoteStream, onIceState) {
  sessionToken = token;
  onRemoteStreamCallback = onRemoteStream;
  onIceConnectionStateCallback = onIceState;

  peerConnection = new RTCPeerConnection({ iceServers: ICE_SERVERS });

  // Add local tracks if available
  if (localStream) {
    localStream.getTracks().forEach((track) => {
      peerConnection.addTrack(track, localStream);
    });
  }

  // Forward ICE candidates to peer via backend
  peerConnection.onicecandidate = ({ candidate }) => {
    if (candidate) {
      sendSignal({
        type: 'ICE_CANDIDATE',
        senderSessionToken: sessionToken,
        candidate: JSON.stringify(candidate),
      });
    }
  };

  // Handle incoming remote stream
  peerConnection.ontrack = (event) => {
    const [remoteStream] = event.streams;
    if (onRemoteStreamCallback) onRemoteStreamCallback(remoteStream);
  };

  // Track ICE connection state for UI
  peerConnection.oniceconnectionstatechange = () => {
    if (onIceConnectionStateCallback) {
      onIceConnectionStateCallback(peerConnection.iceConnectionState);
    }
  };

  return peerConnection;
}

/**
 * Called when this client is designated as the OFFER sender.
 */
export async function startAsOfferer(token, onRemoteStream, onIceState) {
  createPeerConnection(token, onRemoteStream, onIceState);

  const offer = await peerConnection.createOffer();
  await peerConnection.setLocalDescription(offer);

  sendSignal({
    type: 'OFFER',
    senderSessionToken: token,
    sdp: JSON.stringify(offer),
  });
}

/**
 * Called when this client receives an OFFER and must send an ANSWER.
 */
export async function handleOffer(token, offerSdp, onRemoteStream, onIceState) {
  createPeerConnection(token, onRemoteStream, onIceState);

  const offer = JSON.parse(offerSdp);
  await peerConnection.setRemoteDescription(new RTCSessionDescription(offer));

  const answer = await peerConnection.createAnswer();
  await peerConnection.setLocalDescription(answer);

  sendSignal({
    type: 'ANSWER',
    senderSessionToken: token,
    sdp: JSON.stringify(answer),
  });
}

/**
 * Called when this client receives an ANSWER.
 */
export async function handleAnswer(answerSdp) {
  if (!peerConnection) return;
  const answer = JSON.parse(answerSdp);
  await peerConnection.setRemoteDescription(new RTCSessionDescription(answer));
}

/**
 * Called when an ICE candidate is received from the peer.
 */
export async function handleIceCandidate(candidateJson) {
  if (!peerConnection) return;
  try {
    const candidate = JSON.parse(candidateJson);
    await peerConnection.addIceCandidate(new RTCIceCandidate(candidate));
  } catch (e) {
    console.error('Failed to add ICE candidate:', e);
  }
}

/** Mute/unmute the local audio track */
export function setAudioEnabled(enabled) {
  if (!localStream) return;
  localStream.getAudioTracks().forEach((t) => (t.enabled = enabled));
}

/** Enable/disable the local video track */
export function setVideoEnabled(enabled) {
  if (!localStream) return;
  localStream.getVideoTracks().forEach((t) => (t.enabled = enabled));
}

/** Tear down peer connection and stop all local media */
export function closePeerConnection() {
  if (peerConnection) {
    peerConnection.close();
    peerConnection = null;
  }
}

export function stopLocalStream() {
  if (localStream) {
    localStream.getTracks().forEach((t) => t.stop());
    localStream = null;
  }
}

export function getLocalStream() {
  return localStream;
}

File: mobile/lib/services/webrtc_service.dart
// lib/services/webrtc_service.dart
import 'dart:convert';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import '../config/app_config.dart';
import '../models/models.dart';
import 'signaling_service.dart';

typedef StreamCallback = void Function(MediaStream stream);
typedef IceStateCallback = void Function(RTCIceConnectionState state);

class WebRTCService {
  final SignalingService _signaling;

  RTCPeerConnection? _peerConnection;
  MediaStream? _localStream;

  StreamCallback? onRemoteStream;
  IceStateCallback? onIceStateChange;

  WebRTCService(this._signaling);

  MediaStream? get localStream => _localStream;

  static final _iceServers = {
    'iceServers': [
      {'urls': AppConfig.stunUrl},
    ]
  };

  /// Acquires local camera + mic, falls back gracefully.
  Future<Map<String, dynamic>> initLocalStream() async {
    // Try video + audio
    try {
      _localStream = await navigator.mediaDevices.getUserMedia({
        'video': {'facingMode': 'user'},
        'audio': true,
      });
      return {'videoEnabled': true, 'audioEnabled': true, 'stream': _localStream};
    } catch (_) {}

    // Fallback: audio only
    try {
      _localStream = await navigator.mediaDevices.getUserMedia(
          {'video': false, 'audio': true});
      return {'videoEnabled': false, 'audioEnabled': true, 'stream': _localStream};
    } catch (_) {}

    // Receive-only
    return {'videoEnabled': false, 'audioEnabled': false, 'stream': null};
  }

  Future<RTCPeerConnection> _createPeerConnection(String sessionToken) async {
    _peerConnection = await createPeerConnection(_iceServers);

    // Add local tracks
    if (_localStream != null) {
      _localStream!.getTracks().forEach((track) {
        _peerConnection!.addTrack(track, _localStream!);
      });
    }

    // Forward ICE candidates to backend
    _peerConnection!.onIceCandidate = (candidate) {
      if (candidate.candidate != null) {
        _signaling.send(SignalingMessage(
          type: SignalingType.iceCandidate,
          senderSessionToken: sessionToken,
          candidate: jsonEncode(candidate.toMap()),
        ));
      }
    };

    // Handle remote stream
    _peerConnection!.onTrack = (event) {
      if (event.streams.isNotEmpty) {
        onRemoteStream?.call(event.streams.first);
      }
    };

    // Track ICE state
    _peerConnection!.onIceConnectionState = (state) {
      onIceStateChange?.call(state);
    };

    return _peerConnection!;
  }

  /// Creates offer and sends it to the peer via backend.
  Future<void> startAsOfferer(String sessionToken) async {
    await _createPeerConnection(sessionToken);

    final offer = await _peerConnection!.createOffer();
    await _peerConnection!.setLocalDescription(offer);

    _signaling.send(SignalingMessage(
      type: SignalingType.offer,
      senderSessionToken: sessionToken,
      sdp: jsonEncode(offer.toMap()),
    ));
  }

  /// Handles incoming OFFER and replies with ANSWER.
  Future<void> handleOffer(String sessionToken, String offerSdp) async {
    await _createPeerConnection(sessionToken);

    final offerMap = jsonDecode(offerSdp) as Map<String, dynamic>;
    await _peerConnection!.setRemoteDescription(
      RTCSessionDescription(offerMap['sdp'] as String, offerMap['type'] as String),
    );

    final answer = await _peerConnection!.createAnswer();
    await _peerConnection!.setLocalDescription(answer);

    _signaling.send(SignalingMessage(
      type: SignalingType.answer,
      senderSessionToken: sessionToken,
      sdp: jsonEncode(answer.toMap()),
    ));
  }

  Future<void> handleAnswer(String answerSdp) async {
    if (_peerConnection == null) return;
    final answerMap = jsonDecode(answerSdp) as Map<String, dynamic>;
    await _peerConnection!.setRemoteDescription(
      RTCSessionDescription(answerMap['sdp'] as String, answerMap['type'] as String),
    );
  }

  Future<void> handleIceCandidate(String candidateJson) async {
    if (_peerConnection == null) return;
    final map = jsonDecode(candidateJson) as Map<String, dynamic>;
    await _peerConnection!.addCandidate(RTCIceCandidate(
      map['candidate'] as String,
      map['sdpMid'] as String?,
      map['sdpMLineIndex'] as int?,
    ));
  }

  void setAudioEnabled(bool enabled) {
    _localStream?.getAudioTracks().forEach((t) => t.enabled = enabled);
  }

  void setVideoEnabled(bool enabled) {
    _localStream?.getVideoTracks().forEach((t) => t.enabled = enabled);
  }

  void closePeerConnection() {
    _peerConnection?.close();
    _peerConnection = null;
  }

  void stopLocalStream() {
    _localStream?.getTracks().forEach((t) => t.stop());
    _localStream = null;
  }

  void dispose() {
    closePeerConnection();
    stopLocalStream();
  }
}
